File: 프로그래머스/3/43162. 네트워크/네트워크.py
def solution(n, computers):
    parent = [i for i in range(n)]

    def find_parent(x, parent):
        if parent[x] != x:
            parent[x] = find_parent(parent[x], parent)
        return parent[x]

    def make_union(a, b, parent):
        a = find_parent(a, parent)
        b = find_parent(b, parent)

        if a < b:
            parent[b] = a
        else:
            parent[a] = b

    for i in range(n):
        for j in range(n):
            if computers[i][j] == 1 and i != j:
                make_union(i, j, parent)

    for i in range(n):
        parent[i] = find_parent(i, parent)

    return len(set(parent))

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] computers) {
        int answer = 0;
       
        int[] parent=new int[n];
        for(int i=0;i<n;i++){
            parent[i]=i;
        }
        
        for(int i=0;i<n;i++){
            for(int j=0;j<n;j++){
                if(i!=j && computers[i][j]==1){
                    make_union(i,j,parent);
                }
            }
        }
        
        for(int i=0;i<n;i++){
            find_parent(i,parent);
        }
        
        HashSet<Integer> s=new HashSet<>();
        for(int i=0;i<n;i++){
            s.add(parent[i]);
        }
        
        return s.size();
    }
    
    int find_parent(int x, int[]parent){
        if(parent[x]!=x){
            parent[x]=find_parent(parent[x],parent);
        }
        return parent[x];
    }
    
    void make_union(int a, int b, int[]parent){
        a=find_parent(a,parent);
        b=find_parent(b,parent);
        
        if(a<b){
            parent[b]=a;
        }
        else{
            parent[a]=b;
        }
    }
}